clientv3/integration/black_hole_test.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !cluster_proxy
+
+package integration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/integration"
+	"github.com/coreos/etcd/pkg/testutil"
+)
+
+func TestBlackholePutWithoutKeealiveEnabled(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{
+		Size:                 2,
+		GRPCKeepAliveMinTime: 1 * time.Millisecond},
+	) // avoid too_many_pings
+
+	defer clus.Terminate(t)
+
+	ccfg := clientv3.Config{
+		Endpoints:   []string{clus.Members[0].GRPCAddr()},
+		DialTimeout: 1 * time.Second,
+	}
+	cli, err := clientv3.New(ccfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cli.Close()
+
+	// wait for eps[0] to be pinned
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	_, err = cli.Get(ctx, "foo")
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cli.SetEndpoints(clus.Members[0].GRPCAddr(), clus.Members[1].GRPCAddr())
+	clus.Members[0].Blackhole()
+
+	// fail first due to blackhole, retry should succeed
+	// when gRPC supports better retry on non-delivered request, the first put can succeed.
+	ctx, cancel = context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
```

Comment:
This `cancel` is overwritten by line 68?

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	_ = ctx
	defer func() {
		fmt.Println("cancel 1:", cancel)
		cancel()
	}()

	ctx, cancel = context.WithTimeout(context.Background(), time.Second)
	_ = ctx
	defer func() {
		fmt.Println("cancel 2:", cancel)
		cancel()
	}()
}

/*
cancel 2: 0x109a950
cancel 1: 0x109a950
*/
```

Should we move `cancel` to line 65?

Reply:
will fix.